discovery/triton/triton.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2016 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package triton
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"time"
+
+	promlog "github.com/prometheus/common/log"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/config"
+	"golang.org/x/net/context"
+)
+
+const (
+	tritonLabel             = model.MetaLabelPrefix + "triton_"
+	tritonLabelMachineId    = tritonLabel + "machine_id"
+	tritonLabelMachineAlias = tritonLabel + "machine_alias"
+	tritonLabelMachineImage = tritonLabel + "machine_image"
+	tritonLabelServerId     = tritonLabel + "server_id"
+)
+
+type TritonDiscoveryResponse struct {
+	Containers []struct {
+		ServerUUID  string `json:"server_uuid"`
+		VMAlias     string `json:"vm_alias"`
+		VMImageUUID string `json:"vm_image_uuid"`
+		VMUUID      string `json:"vm_uuid"`
+	} `json:"containers"`
+}
+
+// TritonDiscovery periodically performs Triton-SD requests. It implements
+// the TargetProvider interface.
+type TritonDiscovery struct {
+	sdConfig *config.TritonSDConfig
+	client   *http.Client
+	interval time.Duration
+}
+
+// NewTritonDiscovery returns a new TritonDiscovery which periodically refreshes its targets.
+func NewDiscovery(conf *config.TritonSDConfig) *TritonDiscovery {
+	cert, err := tls.LoadX509KeyPair(conf.Cert, conf.Key)
+	if err != nil {
+		log.Fatalln("Unable to load cert", err)
```

Comment:
Exiting just because a single SD failed to start is not the greatest solution.

We have other SDs that can error as well, e.g.:https://github.com/prometheus/prometheus/blob/master/discovery/discovery.go#L98-L105

It's not perfect, but just logging in a single place and continuing with best effort is the better option.